service/history/shard/context_impl.go

```diff
@@ -683,6 +692,91 @@ func (s *ContextImpl) AppendHistoryEvents(
 	return size, err0
 }
 
+func (s *ContextImpl) DeleteWorkflowExecution(
+	key definition.WorkflowKey,
+	branchToken []byte,
+	version int64,
+) error {
+	if err := s.errorByState(); err != nil {
+		return err
+	}
+
+	// do not try to get namespace cache within shard lock
+	namespaceEntry, err := s.GetNamespaceRegistry().GetNamespaceByID(namespace.ID(key.NamespaceID))
+	if err != nil {
+		return err
+	}
+
+	s.wLock()
+	defer s.wUnlock()
+
+	delCurRequest := &persistence.DeleteCurrentWorkflowExecutionRequest{
+		ShardID:     s.shardID,
+		NamespaceID: key.NamespaceID,
+		WorkflowID:  key.WorkflowID,
+		RunID:       key.RunID,
+	}
+	op := func() error {
+		return s.GetExecutionManager().DeleteCurrentWorkflowExecution(delCurRequest)
+	}
+	err = backoff.Retry(op, persistenceOperationRetryPolicy, common.IsPersistenceTransientError)
+	if err != nil {
+		return err
+	}
```

Comment:
Could you please verify that delete won't error out if it could not find the record to delete.  Also if the CAS failed on current row condition, it should not return error. Need to make sure that if current row is deleted, but subsequent delete failed, the retry won't be blocked here.

Reply:
I verified by doubling all the delete calls, and also by trying to delete a nonexistent execution.

For Cass, it seems like delete with IF EXISTS fails if the row doesn't exist, but delete with IF _condition_ succeeds either way (though it tells you whether it found it or not).

Also note that if there's a problem here, then there was a problem before too: the previous code did the same things in (mostly) the same order.